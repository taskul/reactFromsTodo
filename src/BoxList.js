import Box from "./Box";
import BoxForm from "./BoxForm";
import { useState } from "react";
import { v4 as uuidv4 } from 'uuid';

// holds colored boxes and a form for creating colored boxes
const BoxList = () => {
    // sets initial state of boxes with one green box
    const [boxes, setBoxes] = useState([{
        width:100, 
        height:50, 
        color: 'green', 
        id: uuidv4()
    }])

    // function passed to BoxForm for creating of new boxes
    // onSubmit BoxFrom will pass values for new box to addBox function
    // as an object boxObj which then will be deconstructed and 
    // added to a list of boxes
    const addBox = (boxObj) => {
        setBoxes(boxes => [
            ...boxes, {...boxObj, id: uuidv4()}
        ])
    }

    // function that is passed on to child Box component with an id
    // that is assigned to that Box component for removing that component from 
    // a parent component
    const removeBox = (id) => {
        setBoxes(boxes => (
            boxes.filter(box => (box.id !== id))
        ))
    }

    return (
        <div>
            {boxes.map(({width, height, color, id}) => (
                <Box 
                    width={width} 
                    height={height} 
                    color={color} 
                    key={id}
                    removeBox={() => removeBox(id)}
                />
            ))}
            <BoxForm addBox={addBox} />
        </div>
    )
};

export default BoxList;