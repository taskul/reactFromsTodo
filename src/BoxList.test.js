import {render, fireEvent, getByTestId, getByText, within } from '@testing-library/react';
import BoxList from './BoxList';

// smoke test
it('Should render without crashing', function () {
    render(<BoxList />)
})

// snapshot test
it('Should match Snapshot', function () {
    const {asFragment} = render(<BoxList />)
    expect(asFragment()).toMatchSnapshot()
})

// initially should not have width 50px by height 40px red box present in DOM
// then submits form for creating 50px by 40px red box
// then 50px by 40px red box should be present in DOM
// finally clicking on "X" button should remove red box
it('should check form submition', function() {
    const {getByText, queryByTestId, getByLabelText} = render(<BoxList />);

    // checking to make sure box with a testid of red is not present
    expect(queryByTestId('red')).not.toBeInTheDocument();

    // get input fields from the form and a form submittion button
    const width = getByLabelText('Width:');
    const height = getByLabelText('Height:');
    const color = getByLabelText('Color:');
    const btn = getByText('Add')

    // change the values of the input fields
    fireEvent.change(width, {target: {value:'50'}});
    fireEvent.change(height, {target: {value:'40'}});
    fireEvent.change(color, {target:{value:'red'}})
    fireEvent.click(btn);

    // box with testid red should not be present in DOM
    expect(queryByTestId('red')).toBeInTheDocument();

    // remove the box with testid red
    const redBox = queryByTestId('red')
    const removeBtn = within(redBox).getByText('X')

    fireEvent.click(removeBtn)

    // check to make sure box with testid of red is no longer present
    expect(queryByTestId('red')).not.toBeInTheDocument();
})
