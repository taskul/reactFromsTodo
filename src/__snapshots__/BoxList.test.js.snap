// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match Snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="Box"
      data-testid="green"
      style="height: 50px; width: 100px; background-color: green;"
    >
      <button
        class="Box-remove"
      >
        X
      </button>
    </div>
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        placeholder="width"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        placeholder="height"
        type="text"
        value=""
      />
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        placeholder="color"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
